osu.Game/Graphics/UserInterface/BarGraph.cs

```diff
@@ -5,77 +5,233 @@
+        {
+            base.Update();
+
+            if (!bars.Any())
+                return;
+
+            double currentTime = Clock.CurrentTime;
+
+            if (currentTime < animationStartTime + resize_duration)
+            {
+                for (int i = 0; i < bars.Count; i++)
+                {
+                    BarStruct bar = bars[i];
+                    bar.IntermediateValue = Interpolation.ValueAt(currentTime, bar.OldValue, bar.Value, animationStartTime, animationStartTime + resize_duration, easing);
+                    bars[i] = bar;
+                }
+
+                Invalidate(Invalidation.DrawNode);
             }
+            else if (!animationComplete)
+            {
+                for (int i = 0; i < bars.Count; i++)
+                {
+                    BarStruct bar = bars[i];
+                    bar.IntermediateValue = bar.Value;
+                    bars[i] = bar;
+                }
+
+                Invalidate(Invalidation.DrawNode);
+
+                animationComplete = true;
+            }
+        }
+
+        protected override DrawNode CreateDrawNode() => new BarGraphDrawNode(this);
+
+        private class BarGraphDrawNode : DrawNode
+        {
+            public new BarGraph Source => (BarGraph)base.Source;
+
+            public BarGraphDrawNode(BarGraph source)
+                : base(source)
+            {
+            }
+
+            private IShader shader = null!;
+            private Texture texture = null!;
+            private Vector2 drawSize;
+            private BarDirection direction;
+
+            private readonly List<BarStruct> bars = new List<BarStruct>();
+
+            public override void ApplyState()
+            {
+                base.ApplyState();
+
+                shader = Source.shader;
+                texture = Source.texture;
+                drawSize = Source.DrawSize;
+                direction = Source.direction;
+
+                bars.Clear();
+                bars.AddRange(Source.bars);
+            }
+
+            public override void Draw(IRenderer renderer)
+            {
+                base.Draw(renderer);
+
+                if (!bars.Any())
+                    return;
+
+                shader.Bind();
+
+                for (int i = 0; i < bars.Count; i++)
+                {
+                    var bar = bars[i];
+
+                    float barHeight = drawSize.Y * ((direction == BarDirection.TopToBottom || direction == BarDirection.BottomToTop) ? bar.IntermediateValue : bar.ShortSide);
+                    float barWidth = drawSize.X * ((direction == BarDirection.LeftToRight || direction == BarDirection.RightToLeft) ? bar.IntermediateValue : bar.ShortSide);
+
+                    Vector2 topLeft;
+
+                    switch (direction)
+                    {
+                        default:
+                        case BarDirection.LeftToRight:
+                            topLeft = new Vector2(0, i * barHeight);
```

Comment:
Using `i * barHeight` like this implicitly assumes that every bar has the same `barHeight` (i.e. `shortSide`). Which is fine right now since that is true, but (a) it may not be at some point, and - more importantly - (b) if this is to be the case, then why store `shortSide` for every bar separately if it's the same for every one anyways? Seems like a waste of memory.